libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt

```diff
@@ -0,0 +1,201 @@
+package com.mapbox.navigation.core
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.base.common.logger.model.Tag
+import com.mapbox.common.BillingServiceError
+import com.mapbox.common.BillingServiceErrorCode
+import com.mapbox.common.SKUIdentifier
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.core.directions.session.DirectionsSession
+import com.mapbox.navigation.core.trip.session.TripSession
+import com.mapbox.navigation.utils.internal.LoggerProvider
+import com.mapbox.turf.TurfConstants.UNIT_METRES
+import com.mapbox.turf.TurfMeasurement
+import java.util.concurrent.TimeUnit
+
+internal class BillingController(
+    private val accessToken: String,
+    private val navigationSession: NavigationSession,
+    private val tripSession: TripSession
+) {
+
+    private companion object {
+        private val tag = Tag("MbxNavBillingController")
+        private const val MAX_WAYPOINTS_DISTANCE_DIFF_METERS = 100.0
+    }
+
+    private val navigationSessionStateObserver =
+        NavigationSessionStateObserver { navigationSessionState ->
+            // stop any running sessions
+            BillingServiceWrapper.stopBillingSession(SKUIdentifier.NAV2_SES_TRIP)
+
+            if (navigationSessionState != NavigationSession.State.IDLE) {
+                // always trigger an MAU event if a session starts
+                BillingServiceWrapper.triggerBillingEvent(
+                    accessToken,
+                    "", // empty string results in default user agent
+                    SKUIdentifier.NAV2_SES_MAU
+                ) {
+                    handlerError(it)
+                }
+            }
+
+            when (navigationSessionState) {
+                NavigationSession.State.IDLE -> {
+                    // do nothing
+                }
+                NavigationSession.State.FREE_DRIVE -> {
+                    BillingServiceWrapper.beginBillingSession(
+                        SKUIdentifier.NAV2_SES_TRIP,
+                        validity = TimeUnit.HOURS.toMillis(1) // validity of 1hr
+                    )
+                }
+                NavigationSession.State.ACTIVE_GUIDANCE -> {
+                    BillingServiceWrapper.beginBillingSession(
+                        SKUIdentifier.NAV2_SES_TRIP,
+                        validity = 0 // default validity, 12hrs
```

Comment:
Wouldn't be clear to pass `12`hrs (and remove the comment)?

Reply:
We don't want to limit active guidance sessions at all. 12 hours is just a server-side limit that might be lifted at some point. I can remove the `12hrs` note and just indicate that this is max possible duration (whatever that is at a certain point in time).

Reply:
Sounds good 👍 

I'm assuming `0` is the "max" then, right? That's why it's confusing 😅

Reply:
That's right, it is indeed a tiny bit confusing.